trainer/finetune.py

```diff
@@ -0,0 +1,240 @@
+import os
+import tensorboard as tb
+import random
+import nltk
+import sacrebleu
+from absl import app
+from absl import flags
+from trainer import preprocessing
+
+FLAGS = flags.FLAGS
+flags.DEFINE_integer('steps', 6000, "Finetuning training steps.")
+flags.DEFINE_enum("size", "base", ["small", "base", "large", "3B", "11B"], "model size")
+flags.DEFINE_string("name", "default", "name/description of model  version")
+flags.DEFINE_enum("mode", "all", ["train", "evaluate", "all"], "run mode: train, evaluate, or all")
+
+INPUT_LENGTH = 512 #1033 - longest training input for redial
+TARGET_LENGTH = 128 #159 - longest trainin target for redial
+
+warnings.filterwarnings("ignore", category=DeprecationWarning)
+
+BASE_DIR = "gs://e2e_central"
+DATA_DIR = os.path.join(BASE_DIR, "data")
+MODELS_DIR = os.path.join(BASE_DIR, "models")
+
+# Public GCS path for T5 pre-trained model checkpoints
+BASE_PRETRAINED_DIR = "gs://t5-data/pretrained_models"
+PRETRAINED_DIR = ""
+MODEL_DIR = ""
+
+# Locations for reading and writing Redial data
+RD_JSONL_DIR = "gs://e2e_central/data/redial/"
+RD_SPLIT_FNAMES = {
+    "train": "rd-train-formatted.jsonl",
+    "validation": "rd-test-formatted.jsonl"
+}
+rd_counts_path = os.path.join(DATA_DIR, "rd-counts.json")
+rd_tsv_path = {
+    "train": os.path.join(DATA_DIR, "rd-train.tsv"),
+    "validation": os.path.join(DATA_DIR, "rd-validation.tsv")
+}
+
+def tf_verbosity_level(level):
+  """Changes verbosity level."""
+  og_level = tf.logging.get_verbosity()
+  tf.logging.set_verbosity(level)
+  yield
+  tf.logging.set_verbosity(og_level)
+
+def _prediction_file_to_ckpt(path):
+  """Extract the global step from a prediction filename."""
+  return int(path.split("_")[-2])
+
+def load_predictions(task_name):
+  """Loads the most recent predictions in as ([(input, target, pred)], step)."""
+  # Grab the dataset for this task.
+  ds = t5.data.TaskRegistry.get(task_name).get_dataset(
+      split="validation",
+      sequence_length={"inputs": INPUT_LENGTH, "targets": TARGET_LENGTH},
+      shuffle=False)
+
+  # Grab the paths of all logged predictions.
+  prediction_files = tf.io.gfile.glob(
+      os.path.join(
+          MODEL_DIR,
+          "validation_eval/%s_*_predictions" % task_name))
+
+  # Get most recent prediction file by sorting by their step.
+  latest_prediction_file = sorted(
+      prediction_files, key=_prediction_file_to_ckpt)[-1]
+
+  checkpoint_step =_prediction_file_to_ckpt(latest_prediction_file)
+
+  # Collect (inputs, targets, prediction) from the dataset and predictions file
+  results = []
+  with tf.io.gfile.GFile(latest_prediction_file) as preds:
+    for ex, pred in zip(tfds.as_numpy(ds), preds):
+      results.append((tf.compat.as_text(ex["inputs_plaintext"]),
+                      tf.compat.as_text(ex["targets_plaintext"]),
+                      pred.strip()))
+  return (results, checkpoint_step)
+
+def save_metrics(task_name):
+  """Prints and saves metrics for the most recent checkpoint."""
```

Comment:
Extend comment to mention:
 - which metrics are being compute
 - Any relevant details such as lowercasing or not, replacing movie names or not, etc